### ISSUE TITLE
question: get action payload in lifecycle hooks

### ISSUE THREAD

**Issue by razvanioan** [vuex-orm/vuex-orm#172]
Is there a way to obtain the action payload inside a lifecycle hook ?

For example what's being requested to update in a beforeUpdate hook ?

I see the `record` is having the new values in the update request, but do I need to check all fields in order to see what's being changed ?

**Comment by kiaking** [comment]
Yes! You can access the current value by obtaining the getter context. But I think having access to the previous state is a nice idea. However, I think we should be documenting that namespace issue as well.

> I see that now we can add hooks inside Modules (and you plan to support them directly on Models too), but is there a way to attache / detach them from code, outside Modulde / Model ?

Not currently. Since right now, lifecycle hook is retrieved each time the lifecycle is called. I think having a way to not execute some hook is a good idea. Maybe something like this?

```js
store.dispatch('entities/users/update', {
  where: 1,
  data: { ... },
  hooks: false // Don't execute hook on this call. 
})
```

**Comment by razvanioan** [comment]
Yeah, actually getting the record like I did, you have access on it's state before the requested update.

On the second parameter that you set, it's the record with updated values, that you can check and allow or not.

In my check I needed to know if the field ’status’ value has been changed in the current request or not.

That's why it should be nice to have in there also the payload (data requested to be updated)

This is handy too in another particular case that of course I hit it :)

- Inside ’beforeUpdate’ I make a check for record state locked or not (now using the old record and the updated one)
- if it's locked I return false to cancel the change
- I send a notification to the user about the reason why he cannot change
- and I let him choose if he wants to unlock it, it's actually a custom confirmation using a ’Promise()’
- and in case he confirms the unlock, after I do what's needed in order to unlock it, **I would have requested his original change again**

But in this case, I really need to know what his change payload it was

Of course there's still a possibility to compare the two record states, and extract the info from there, but it's a little overkill, especially if we can have access to the payload inside the change hook easier ;)

Or another idea, if the before lifecycle hooks would return promisses in order to return async false or not ;)

What do you think ?

Thx,

**Comment by razvanioan** [comment]
Regarding async actions, I mean a user confirmation

It's a good ideea to wait for a Promise(resolve, reject) which can be either an API call or more usefull a user confirmation from UI, in order to continue with the requested change or not

**Comment by kiaking** [comment]
> Having also updated payload separately would be easier to see what's requested, thus you will need to compare two objects (state record before and after change) in order to see that

That is true, but it is more flexible so I think it's a way to go for a first step : )

> Regarding async actions, I mean a user confirmation

It is a pretty cool idea. Not sure how the API would look like though...
